services/src/main/java/org/keycloak/social/microsoft/MicrosoftIdentityProvider.java

```diff
@@ -71,58 +65,62 @@ protected String getProfileEndpointForValidation(EventBuilder event) {
-        return SimpleHttp.doGet(URL, session);
-    }
-
     @Override
     protected BrokeredIdentityContext doGetFederatedIdentity(String accessToken) {
         try {
-            String URL = PROFILE_URL + "?access_token=" + URLEncoder.encode(accessToken, "UTF-8");
-            if (log.isDebugEnabled()) {
-                log.debug("Microsoft Live user profile request to: " + URL);
-            }
-            JsonNode profile = SimpleHttp.doGet(URL, session).asJson();
-
+            JsonNode profile = SimpleHttp.doGet(PROFILE_URL, session).auth(accessToken).asJson();
             return extractIdentityFromProfile(null, profile);
         } catch (Exception e) {
-            throw new IdentityBrokerException("Could not obtain user profile from Microsoft Live ID.", e);
+            throw new IdentityBrokerException("Could not obtain user profile from Microsoft Graph", e);
         }
     }
 
     @Override
     protected BrokeredIdentityContext extractIdentityFromProfile(EventBuilder event, JsonNode profile) {
         String id = getJsonProperty(profile, "id");
-
-        String email = null;
-        if (profile.has("emails")) {
-            email = getJsonProperty(profile.get("emails"), "preferred");
-        }
-
         BrokeredIdentityContext user = new BrokeredIdentityContext(id);
 
+        String email = getJsonProperty(profile, "mail");
+        if (email == null && profile.has("userPrincipalName")) {
+            String username = getJsonProperty(profile, "userPrincipalName");
+            if (this.isValidEmail(username)) {
+                email = username;
+            }
+        }
         user.setUsername(email != null ? email : id);
-        user.setFirstName(getJsonProperty(profile, "first_name"));
-        user.setLastName(getJsonProperty(profile, "last_name"));
+        user.setFirstName(getJsonProperty(profile, "givenName"));
+        user.setLastName(getJsonProperty(profile, "surname"));
         if (email != null)
             user.setEmail(email);
         user.setIdpConfig(getConfig());
         user.setIdp(this);
 
         AbstractJsonUserAttributeMapper.storeUserProfileForMapper(user, profile, getConfig().getAlias());
-
         return user;
     }
 
     @Override
     protected String getDefaultScopes() {
         return DEFAULT_SCOPE;
     }
+
+    /**
+     * Verifies if the supplied string is a valid e-mail address using the regex specified in the
+     * <a href="https://www.owasp.org/index.php/OWASP_Validation_Regex_Repository">OWASP Validation Regex repository</a>
+     *
+     * @param email the string to be verified.
+     * @return {@code true} if the supplied string is a valid e-mail address; {@code false} otherwise.
+     */
+    public boolean isValidEmail(String email)
+    {
+        String emailRegex = "^[a-zA-Z0-9_+&*-]+(?:\\."+
+                "[a-zA-Z0-9_+&*-]+)*@" +
+                "(?:[a-zA-Z0-9-]+\\.)+[a-z" +
+                "A-Z]{2,7}$";
+
+        Pattern pat = Pattern.compile(emailRegex);
```

Comment:
I'd even prefer to go with current Validation.isEmailValid since it is less invalid. Domain names can be UTF-8 currently, not [a-zA-Z0-9-], and there is no limit for TLD. See http://www.iana.org/domains/root/db

Reply:
@hmlnarik +1. That will be easier. From quick googling, it seems that owasp validation is too restrictive. And there doesn't seem to be any nice regex, which matches all the valid rules for the email address